alternator/conditions.cc

```diff
@@ -122,6 +122,35 @@ static bool check_BEGINS_WITH(const rjson::value& v1, const rjson::value& v2) {
     return val1.substr(0, val2.size()) == val2;
 }
 
+// Verify that an Expect comparison operator holds between got and expected value.
+static bool verify_expected_comparison_operator(
+        const rjson::value* got,
+        const rjson::value& comparison_operator,
+        const rjson::value& expected) {
+    // TODO: when IN, BETWEEN, IS_NULL, or NOT_NULL are supported, this will depend on operator:
```

Comment:
The reason I went with a switch statement was to move all these operator dependencies into single blocks. Right now, it will end up with a couple of branches, right? The structure I was trying to avoid was:

if (in or between) {
    assert that expected.Size() == 2;
} else {
    assert that expected.Size() == 1;
}
(...)
if (!got) {
    special case for NE; otherwise return false;
}
(...)
if (in or between) {
    use expected[0] and expected[1]
} else {
   use expected[0] only
}

I think that the unordered_map approach fits well now, because we only implemented similar operators with a single parameter, but later it will end up in using both a map and a bunch of ifs, compared to a single visit-like switch statement (which, fair enough, comes with a bit of boilerplate code every time). Then, when implementing another operator, it will be enough to add a single isolated switch case.

But, it's not a blocker for me, semantically the change looks valid.

Reply:
> The reason I went with a switch statement was to move all these operator dependencies into single blocks.

One problem with that is that there's considerable code repetition across blocks.  That will quickly become unwieldy as we add support for the remaining operators.  That makes both coding and maintenance more laborious; also bugs like [this](https://github.com/scylladb/scylla/pull/5108#issuecomment-535657340) sneak in all too easily.

What would you say to something like this:

```c++
switch (optype) {
  case BETWEEN:
    validate_size(expected, 2); // probably pass in comparison_operator, too, for error formating
    return got ? check_between(*got, expected[0], expected[1]) : false;
  case IN:
    // don't validate_size
    return got ? check_in(*got, expected) : false;
  case NE:
    validate_size(expected, 1);
    return got ? check_ne(*got, expected[0]) : true;
  default:
    validate_size(expected, 1);
    return got ? (*checks)(got, expected[0]) : false;
}
```
We could tighten it even more by moving the `got` check into `check_*` functions, but then we'd have to change all their other call sites outside `conditions.cc`, of which there are several.

Reply:
That one looks very neat. I get the point about code repetitions leading to errors too, I'm fine with the patch as is.

Reply:
OK, I'll refactor it that way when I add more operators.